Gab/days before maturity (#678)

GabrielSandoval opened the pull request:
Old /plantings page:
![plantings_old](https://cloud.githubusercontent.com/assets/6015897/8326510/3966f7be-1a94-11e5-9799-a3b8db625368.png)

New /plantings page:
![plantings_new](https://cloud.githubusercontent.com/assets/6015897/8326521/4911b960-1a94-11e5-9fce-f2d9e653072c.png)

Hi
I added this feature and redesigned the plantings' index page (see photo above). I added a days_before_planting column on the plantings table. I based the calculations on the previous plantings' start and end date.

Also, I added the days_before_maturity attribute on the show page.
![screen shot 2015-06-25 at 12 33 32 pm](https://cloud.githubusercontent.com/assets/6015897/8346954/a027a208-1b36-11e5-94b4-0b34b4ed433d.png)

Gabriel Sandoval,
AELOGICA Intern

maco commented:
When you said "I added a days_before_planting column" did you mean days
before maturity? I'm not seeing anything about days until planting

On Thu, Jun 25, 2015 at 7:01 PM Gabriel <notifications@github.com> wrote:

> Old /plantings page:
> [image: plantings_old]
> <https://cloud.githubusercontent.com/assets/6015897/8326510/3966f7be-1a94-11e5-9799-a3b8db625368.png>
>
> New /plantings page:
> [image: plantings_new]
> <https://cloud.githubusercontent.com/assets/6015897/8326521/4911b960-1a94-11e5-9fce-f2d9e653072c.png>
>
> Hi
> I added this feature and redesigned the plantings index page (see photo
> above). I added a days_before_planting column on the plantings table. I
> based the calculations on the previous plantings' start and end date.
>
> Also, I added the days_before_maturity attribute on the show page.
> [image: screen shot 2015-06-25 at 12 33 32 pm]
> <https://cloud.githubusercontent.com/assets/6015897/8346954/a027a208-1b36-11e5-94b4-0b34b4ed433d.png>
>
> Gabriel Sandoval,
> AELOGICA Intern
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/Growstuff/growstuff/pull/758
> Commit Summary
>
>    - Reply to notification action
>    - Removed excessive example from notification controller specs
>    - Adding myself to contributors
>    - Added the days before maturity feature #678
>
> File Changes
>
>    - *M* CONTRIBUTORS.md
>    <https://github.com/Growstuff/growstuff/pull/758/files#diff-0> (2)
>    - *M* app/controllers/notifications_controller.rb
>    <https://github.com/Growstuff/growstuff/pull/758/files#diff-1> (15)
>    - *M* app/controllers/plantings_controller.rb
>    <https://github.com/Growstuff/growstuff/pull/758/files#diff-2> (11)
>    - *M* app/helpers/notifications_helper.rb
>    <https://github.com/Growstuff/growstuff/pull/758/files#diff-3> (7)
>    - *M* app/models/ability.rb
>    <https://github.com/Growstuff/growstuff/pull/758/files#diff-4> (1)
>    - *M* app/models/planting.rb
>    <https://github.com/Growstuff/growstuff/pull/758/files#diff-5> (15)
>    - *A* app/views/notifications/_notification.html.haml
>    <https://github.com/Growstuff/growstuff/pull/758/files#diff-6> (13)
>    - *A* app/views/notifications/reply.html.haml
>    <https://github.com/Growstuff/growstuff/pull/758/files#diff-7> (6)
>    - *M* app/views/notifications/show.html.haml
>    <https://github.com/Growstuff/growstuff/pull/758/files#diff-8> (14)
>    - *A* app/views/plantings/_planting_progress.html.haml
>    <https://github.com/Growstuff/growstuff/pull/758/files#diff-9> (17)
>    - *M* app/views/plantings/index.html.haml
>    <https://github.com/Growstuff/growstuff/pull/758/files#diff-10> (91)
>    - *M* app/views/plantings/show.html.haml
>    <https://github.com/Growstuff/growstuff/pull/758/files#diff-11> (10)
>    - *M* config/routes.rb
>    <https://github.com/Growstuff/growstuff/pull/758/files#diff-12> (4)
>    - *A*
>    db/migrate/20150625224805_add_days_before_maturity_to_plantings.rb
>    <https://github.com/Growstuff/growstuff/pull/758/files#diff-13> (5)
>    - *M* spec/controllers/notifications_controller_spec.rb
>    <https://github.com/Growstuff/growstuff/pull/758/files#diff-14> (12)
>    - *A* spec/features/notifications_spec.rb
>    <https://github.com/Growstuff/growstuff/pull/758/files#diff-15> (25)
>    - *M* spec/helpers/notifications_helper_spec.rb
>    <https://github.com/Growstuff/growstuff/pull/758/files#diff-16> (5)
>
> Patch Links:
>
>    - https://github.com/Growstuff/growstuff/pull/758.patch
>    - https://github.com/Growstuff/growstuff/pull/758.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Growstuff/growstuff/pull/758>.
>

GabrielSandoval commented:
Oh, I'm so sorry. Yes, I meant 'days_before_maturity'.

GabrielSandoval commented:
Hi @maco,

I'm wondering if you're interested to this feature. I'd like to know if I should push on to refactoring and making feature tests about this one.

Thanks.

Gabriel Sandoval,
AELOGICA Intern

pozorvlak commented:
Wow, looks great! "Yes please" to the refactoring and testing, and please also add yourself to CONTRIBUTORS.md.

GabrielSandoval commented:
Hi @pozorvlak,

I'm done with adding tests for this and I also fixed some minor bugs. Just tell me if there's anything more needed to be done.

Thanks!

Gabriel Sandoval,
AELOGICA Intern

pozorvlak commented on app/views/plantings/_planting_progress.html.haml:
```diff
@@ -0,0 +1,18 @@
+- if DateTime.now.to_date < planting.planted_at
+	= "Progress: 0% - not planted yet"
+	.progress
```
I don't like the repetition here - could you factor out the progress bar into another partial? [Here's an example of how to pass variables when rendering partials](https://github.com/Growstuff/growstuff/blob/dev/app/views/alternate_names/show.html.haml#L3).

pozorvlak commented on app/views/plantings/index.html.haml:
```diff
@@ -22,40 +22,71 @@
+            %li
+              Crop name:
+              =link_to planting.crop.name, planting.crop
+            - if planting.finished?
```
There's a lot of moderately complex logic shared between here and your changes in  app/views/plantings/show.html.haml: it would be better if you could factor out the common logic into a helper function, or perhaps a method on the Planting model.

pozorvlak commented on app/views/plantings/index.html.haml:
```diff
@@ -22,40 +22,71 @@
-        %td= planting.quantity
-        %td= planting.planted_at
+        %td= "##{index+1}"
+        %td= link_to image_tag((planting.default_photo ?  planting.default_photo.thumbnail_url : 'placeholder_150.png'), :alt => planting.crop_id, :class => 'img'), planting
```
I think the code in the body of this loop should be moved into plantings/_thumbnail.html.haml, where it can replace the rather dull planting-summary that we're currently using in Garden detail views. Then replace the loop body here with a call to the thumbnail partial :-)

pozorvlak commented:
Great stuff - we're definitely getting there! Just a couple of refactorings as described above, and I think this will be good to merge.

GabrielSandoval commented:
Hi @pozorvlak,

I finished the refactoring on the index, show, thumbnails and progress views. I created a helper file for plantings. Also, as requested, I replaced the old dull plantings/_thumbnail view with the new one.

I had some issues regarding the test about the plantings/_thumbnail view but since those tests are deprecated, I decided to comment them out to get rid of errors.

pozorvlak commented:
When I said "the body of the loop", I meant the block *inside* the `each` - all our other thumbnail partials only show one object, not a list of objects. Also, I belatedly see you've used HTML tables rather than Bootstrap CSS classes - I don't know if that's likely to cause problems. @cesy @maco @Skud? It would be better if you could do it in a way consistent with the rest of the site's code, but I don't know how hard that would be.

Committing commented-out code is a [really bad idea](http://programmers.stackexchange.com/questions/45378/is-commented-out-code-really-always-bad) - if we need old code back again, we have version control :-) In this case it's unsurprising the tests failed, since you were changing the behaviour of that partial quite dramatically. Since your new code is covered by feature tests, I think it's fine to just delete the thumbnail view spec file.

> I'm looking forward to this getting merged.

Me too - I'm sorry this is taking so long! I think it's been a valuable process, though.

I'll be away for the next few days in a region without phone or Internet access, so won't be able to look at this again until Monday. Hopefully one of the other committers will be around in the meantime :-)

maco commented:
I'll keep an eye out while you're gone.

The table display is going to be bad for screen readers, especially because
you have a ul in a td several times. The entire ul will be read top to
bottom instead of matching up label and value. I would use a definition
list (dl) and some css to display the planting date etc.

On Fri, Jul 3, 2015, 4:48 AM pozorvlak <notifications@github.com> wrote:

> When I said "the body of the loop", I meant the block *inside* the each -
> all our other thumbnail partials only show one object, not a list of
> objects. Also, I belatedly see you've used HTML tables rather than
> Bootstrap CSS classes - I don't know if that's likely to cause problems.
> @cesy <https://github.com/cesy> @maco <https://github.com/maco> @Skud
> <https://github.com/Skud>? It would be better if you could do it in a way
> consistent with the rest of the site's code, but I don't know how hard that
> would be.
>
> Committing commented-out code is a really bad idea
> <http://programmers.stackexchange.com/questions/45378/is-commented-out-code-really-always-bad>
> - if we need old code back again, we have version control :-) In this case
> it's unsurprising the tests failed, since you were changing the behaviour
> of that partial quite dramatically. Since your new code is covered by
> feature tests, I think it's fine to just delete the thumbnail view spec
> file.
>
> I'm looking forward to this getting merged.
>
> Me too - I'm sorry this is taking so long! I think it's been a valuable
> process, though.
>
> I'll be away for the next few days in a region without phone or Internet
> access, so won't be able to look at this again until Monday. Hopefully one
> of the other committers will be around in the meantime :-)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Growstuff/growstuff/pull/758#issuecomment-118279351>.
>

GabrielSandoval commented:
Thanks for the tip @maco and @pozorvlak !  I'll be working on it.

GabrielSandoval commented:
Hi @maco , @pozorvlak ,

I've deleted the thumbnail view spec file, changed the ul into dl, made the thumbnail partial pass only a single object, and converted the html tables into bootstrap css classes. All tests passed.

I'll be waiting for your response. Thanks!

GabrielSandoval commented:
Thanks @maco !
I'll be looking for more issues to work on. :)

pozorvlak commented:
Hurrah - good work!

GabrielSandoval commented:
Thanks too @pozorvlak ! I've learned so much from you guys!
